blog/_posts/2023-04-18-integrating-scalajs-with-vite.md

```diff
@@ -0,0 +1,127 @@
+---
+layout: blog-detail
+post-type: blog
+by: Sébastien Doeraene
+title: Integrating Scala.js with Vite
+description: Using the new vite-plugin-scalajs to better integrate Scala.js with the Vite build tool.
+---
+
+One of the cornerstones of Scala.js is its ability to interoperate with JavaScript, in order to leverage its massive ecosystem.
+At the language level, we have good foundations and good tooling (namely [ScalablyTyped](https://scalablytyped.org/)) to reach out to JavaScript libraries.
+However, at the tooling level, Scala.js remains centered around Scala tooling.
+While this is good for the *coding* part of development, it is not so great for iterative feedback loops with the browser, or integration of other front-end assets.
+
+The JavaScript ecosystem features a lot of great tooling in that respect.
+[Vite](https://vitejs.dev/), in particular, is an interactive-first build tool for front-end development.
+It provides live-reloading with fast refresh cycles.
+Leveraging that tooling in the context of Scala.js would improve our ability to rapidly develop Scala.js applications.
+
+In this post, we introduce [`vite-plugin-scalajs`](https://github.com/scala-js/vite-plugin-scalajs), a Vite plugin to integrate a Scala.js codebase into the Vite development pipeline.
+In addition, we present a new tutorial series for Scala.js that integrates [Laminar](https://laminar.dev/) and ScalablyTyped with Vite.
+
+## Past approach: scalajs-bundler
+
+Several years ago, the Scala Center first developed [scalajs-bundler](https://scalacenter.github.io/scalajs-bundler/index.html).
+Broadly, its goal was similar to what we present today: to integrate Scala.js with JavaScript tooling, in particular [npm](https://www.npmjs.com/) and [webpack](https://webpack.js.org/).
+
+scalajs-bundler is an sbt plugin that provides settings and tasks to orchestrate npm and webpack invocations.
+The main premise was that Scala.js developers would be using Scala tooling first.
+They would only want the JavaScript tooling as far as basic dependency management and bundling requirements were concerned.
+With that premise, scalajs-bundler abstracts away npm and webpack to some extent, and is intended to be used from sbt.
+
+While that approach works well for a while, it does not scale well to deeper front-end development requirements.
+Soon enough, scalajs-bundler had to add support for custom webpack configuration files.
+Unfortunately, using that feature meant that the webpack configuration was actually spread across sbt- and webpack-style configuration.
+
+As time passed, best practices for Scala.js development evolved.
```

Comment:
Also, as time passed, module splitting was implemented, which opened the possibility to leverage tools like Vite. At the time scalajs-bundler was created, Scala.js was only able to emit a single huge JS module, and we had to implement hacks to provide a fast refresh cycle. scalajs-bundler was used mainly to produce the final, minified, browser-compatible, bundle.

I think it would make sense to mention that module splitting support in Scala.js opened the door to better integration with JS tooling.